src/wbase/Task.cc

```diff
@@ -125,39 +105,73 @@ std::atomic<uint32_t> taskSequence{0};
 /// available to define the action to take when this task is run, so
 /// Command::setFunc() is used set the action later. This is why
 /// the util::CommandThreadPool is not called here.
-Task::Task(TaskMsgPtr const& t, std::string const& query, int fragmentNumber,
-           std::shared_ptr<SendChannelShared> const& sc)
-        : msg(t),
+Task::Task(TaskMsgPtr const& t, int fragmentNumber, std::shared_ptr<UserQueryInfo> const& userQueryInfo,
+           size_t templateId, int subchunkId, std::shared_ptr<SendChannelShared> const& sc)
+        : _userQueryInfo(userQueryInfo),
           _sendChannel(sc),
           _tSeq(++taskSequence),
           _qId(t->queryid()),
+          _templateId(templateId),
+          _hasChunkId(t->has_chunkid()),
+          _chunkId(t->has_chunkid() ? t->chunkid() : -1),
+          _subchunkId(subchunkId),
           _jId(t->jobid()),
           _attemptCount(t->attemptcount()),
-          _idStr(makeIdStr()),
-          _queryString(query),
-          _queryFragmentNum(fragmentNumber) {
-    hash = hashTaskMsg(*t);
-
+          _queryFragmentNum(fragmentNumber),
+          _fragmentHasSubchunks(t->fragment(fragmentNumber).has_subchunks()),
+          _session(t->has_session() ? t->session() : -1),
+          _hasDb(t->has_db()),
+          _db(t->has_db() ? t->db() : ""),
+          _protocol(t->has_protocol() ? t->protocol() : -1),
+          _czarId(t->has_czarid() ? t->czarid() : -1) {
     if (t->has_user()) {
         user = t->user();
     } else {
         user = defaultUser;
     }
-    timestr[0] = '\0';
 
     allIds.add(std::to_string(_qId) + "_" + std::to_string(_jId));
     LOGS(_log, LOG_LVL_DEBUG,
          "Task(...) "
                  << "this=" << this << " : " << allIds);
 
     // Determine which major tables this task will use.
-    int const size = msg->scantable_size();
+    int const size = t->scantable_size();
     for (int j = 0; j < size; ++j) {
-        _scanInfo.infoTables.push_back(proto::ScanTableInfo(msg->scantable(j)));
+        _scanInfo.infoTables.push_back(proto::ScanTableInfo(t->scantable(j)));
     }
-    _scanInfo.scanRating = msg->scanpriority();
+    _scanInfo.scanRating = t->scanpriority();
     _scanInfo.sortTablesSlowestFirst();
-    _scanInteractive = msg->scaninteractive();
+    _scanInteractive = t->scaninteractive();
+
+    // Create sets and vectors for 'aquiring' subchunk temporary tables.
+    proto::TaskMsg_Fragment const& fragment(t->fragment(_queryFragmentNum));
+    if (!_fragmentHasSubchunks) {
+        /// &&& ??? Why aquire anything if there are no subchunks in the fragment?
+        /// &&& ??? All tasks for this user query probably use the same tables
+        for (auto const& scanTbl : t->scantable()) {
+            _dbTbls.emplace(scanTbl.db(), scanTbl.table());
+        }
+        assert(t->has_db());
```

Comment:
I think this assertion has to be done much earlier in this constructor regardless of the state of `_fragmentHasSubchunks`.

Reply:
The assert is a holdover from the previous code and is probably a really bad way to go about that as it will kill the worker and cause massive trouble. If `has_db()` is false, is will cause `QueryRunner` problems that should get an error sent back to the czar and maybe give the user a decent error message.